annif/backend/xtransformer.py

```diff
@@ -0,0 +1,252 @@
+"""Annif backend using the transformer variant of pecos."""
+
+import logging
+import os.path as osp
+import sys
+from typing import Any
+
+import numpy as np
+import scipy.sparse as sp
+from pecos.utils.featurization.text.preprocess import Preprocessor
+from pecos.xmc.xtransformer import matcher, model
+from pecos.xmc.xtransformer.model import XTransformer
+from pecos.xmc.xtransformer.module import MLProblemWithText
+
+from annif.corpus.document import DocumentCorpus
+from annif.exception import NotInitializedException, NotSupportedException
+from annif.suggestion import SubjectSuggestion, SuggestionBatch
+from annif.util import (
+    apply_param_parse_config,
+    atomic_save,
+    atomic_save_folder,
+    boolean,
+)
+
+from . import backend, mixins
+
+
+class XTransformerBackend(mixins.TfidfVectorizerMixin, backend.AnnifBackend):
+    """XTransformer based backend for Annif"""
+
+    name = "xtransformer"
+    needs_subject_index = True
+
+    _model = None
+
+    train_X_file = "xtransformer-train-X.npz"
+    train_y_file = "xtransformer-train-y.npz"
+    train_txt_file = "xtransformer-train-raw.txt"
+    model_folder = "xtransformer-model"
+
+    PARAM_CONFIG = {
+        "min_df": int,
+        "ngram": int,
+        "fix_clustering": boolean,
+        "nr_splits": int,
+        "min_codes": int,
+        "max_leaf_size": int,
+        "imbalanced_ratio": float,
+        "imbalanced_depth": int,
+        "max_match_clusters": int,
+        "do_fine_tune": boolean,
+        "model_shortcut": str,
+        "beam_size": int,
+        "limit": int,
+        "post_processor": str,
+        "negative_sampling": str,
+        "ensemble_method": str,
+        "threshold": float,
+        "loss_function": str,
+        "truncate_length": int,
+        "hidden_droput_prob": float,
+        "batch_size": int,
+        "gradient_accumulation_steps": int,
+        "learning_rate": float,
+        "weight_decay": float,
+        "adam_epsilon": float,
+        "num_train_epochs": int,
+        "max_steps": int,
+        "lr_schedule": str,
+        "warmup_steps": int,
+        "logging_steps": int,
+        "save_steps": int,
+        "max_active_matching_labels": int,
+        "max_num_labels_in_gpu": int,
+        "use_gpu": boolean,
+        "bootstrap_model": str,
```

Comment:
Adding the following lines to `PARAM_CONFIG` would allow us to make use of the hyperparamters `Cp` and `Cn` in the project configurations:
```
"Cn": float,
"Cp": float,
```
And similarly for the dict `DEFAULT_PARAMETERS`:
```
"Cn": 1.0,
"Cp": 1.0,
```